### ISSUE TITLE
could not find `Cargo.toml` in `/` or any parent directory

### ISSUE THREAD

**Issue by norru** [eclipse/corrosion#233]
This error is popping up randomly when ctrl+clicking on symbols resolving to a `.cargo` lib

![selection_155](https://user-images.githubusercontent.com/889291/53882425-37b77300-400e-11e9-98be-8e937a67c3ba.png)

- I *do* have `.cargo` imported as a project in the workspaces.
- Most of the time, the operation correctly loads the source file in `.cargo` and *does not* produce an error

**Comment by mickaelistria** [comment]
When this happens, can you please share the interesting parts of the language server logs (you can enable log to file in the Language Servers > Logs preference page) ?
Seeing the messages that are exchanged will help to figure out a pattern to reproduce it.

**Comment by norru** [comment]
Yes, I've got LSP4E constantly logging to a log file. I probably still have the relevant event to the error above, but probably not worth the effort to dig it out :) Will do as soon as it happens again.

**Comment by norru** [comment]
```
{"jsonrpc":"2.0","id":"113","method":"textDocument/definition","params":{"textDocument":{"uri":"file:///[MY_SOURCE_FILE].rs"},"position":{"line":189,"character":17}}}
[t=1551896704617] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","id":"113","result":[{"uri":"file:///home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/collections/hash/map.rs","range":{"start":{"line":408,"character":11},"
[t=1551896704827] LSP4E to org.eclipse.corrosion.rls:
Content-Length: 219


[t=1551896704827] LSP4E to org.eclipse.corrosion.rls:
{"jsonrpc":"2.0","id":"114","method":"textDocument/definition","params":{"textDocument":{"uri":"file:///[MY_SOURCE_FILE].rs"},"position":{"line":189,"character":17}}}
[t=1551896704828] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","id":"114","result":[{"uri":"file:///home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/collections/hash/map.rs","range":{"start":{"line":408,"character":11},"
[t=1551896704879] LSP4E to org.eclipse.corrosion.rls:
Content-Length: 226


[t=1551896704880] LSP4E to org.eclipse.corrosion.rls:
{"jsonrpc":"2.0","id":"115","method":"textDocument/documentHighlight","params":{"textDocument":{"uri":"file:///[MY_SOURCE_FILE].rs"},"position":{"line":189,"character"
[t=1551896704880] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","id":"115","result":[]}
[t=1551896705085] LSP4E to org.eclipse.corrosion.rls:
Content-Length: 1164


[t=1551896705086] LSP4E to org.eclipse.corrosion.rls:
{"jsonrpc":"2.0","id":"1","method":"initialize","params":{"processId":23103,"rootUri":"file:///","initializationOptions":{"settings":{"rust":{"clippy_preference":"on"}}},"capabilities":{"workspace":{"applyEdit":true,"wor
[t=1551896705093] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","id":"1","result":{"capabilities":{"textDocumentSync":2,"hoverProvider":true,"completionProvider":{"resolveProvider":true,"triggerCharacters":[".",":"]},"definitionProvider":true,"implementationProvider"
[t=1551896705186] LSP4E to org.eclipse.corrosion.rls:
Content-Length: 52


[t=1551896705187] LSP4E to org.eclipse.corrosion.rls:
{"jsonrpc":"2.0","method":"initialized","params":{}}
[t=1551896705187] LSP4E to org.eclipse.corrosion.rls:
Content-Length: 222


[t=1551896705188] LSP4E to org.eclipse.corrosion.rls:
{"jsonrpc":"2.0","id":"2","method":"textDocument/codeLens","params":{"textDocument":{"uri":"file:///home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/collections/hash/map.rs"}
[t=1551896705189] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"id":"progress_1","title":"Building"}}
[t=1551896705189] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"done":true,"id":"progress_1","title":"Building"}}
[t=1551896705190] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"id":"progress_0","title":"Indexing"}}
[t=1551896705190] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/showMessage","params":{"message":"could not find `Cargo.toml` in `/` or any parent directory","type":1}}
[t=1551896705210] LSP4E to org.eclipse.corrosion.rls:
Content-Length: 157864
```

**Comment by norru** [comment]
The IDE is now frozen.

**Comment by norru** [comment]
Perhaps IDE is not recovering after a RLS crash?

**Comment by norru** [comment]
The IDE is actually freezing while attempting to *close* the `map.rs` file.

**Comment by norru** [comment]
Lots of `rls` instances running after closing the IDE:

```
norru     9502     1  0 16:34 ?        00:00:55 /home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rls
norru     9541     1  0 16:34 ?        00:00:55 /home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rls
norru    10031     1  0 13:14 ?        00:02:09 /home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rls
norru    16351     1  0 16:53 ?        00:00:49 /home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rls
norru    25650     1  1 18:25 ?        00:00:08 /home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rls
norru    26077     1  5 18:33 ?        00:00:05 /home/norru/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rls
```

**Comment by norru** [comment]
Stuck here now. Restart doesn't fix it (gets stuck again).

![selection_156](https://user-images.githubusercontent.com/889291/53904816-dfe53000-403e-11e9-862e-8a78f1c723a5.png)

**Comment by norru** [comment]
Manually removing that file fixes it. Unfortunately this is not a viable workaround.

**Comment by norru** [comment]
Spoke too soon. IDE stuck again on closing the `map.rs` editor.

**Comment by norru** [comment]
Nothing wrong in the LSP4E log
```
{"jsonrpc":"2.0","method":"window/progress","params":{"id":"progress_1","message":"metashape_doc_parser cfg(test)","title":"Building"}}
[t=1551897613816] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"id":"progress_1","message":"parallax_testbed","title":"Building"}}
[t=1551897619021] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"done":true,"id":"progress_1","title":"Building"}}
[t=1551897619022] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"id":"progress_0","title":"Indexing"}}
[t=1551897622218] org.eclipse.corrosion.rls to LSP4E:
{"jsonrpc":"2.0","method":"window/progress","params":{"done":true,"id":"progress_0","title":"Indexing"}}
```

**Comment by norru** [comment]
```
   java.lang.Thread.State: BLOCKED (on object monitor)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:67)
        - waiting to lock <0x00000003c9a551f8> (a java.lang.Object)
        at org.eclipse.lsp4e.LanguageServerWrapper.lambda$1(LanguageServerWrapper.java:213)
        at org.eclipse.lsp4e.LanguageServerWrapper$$Lambda$465/493801605.consume(Unknown Source)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.notify(RemoteEndpoint.java:126)
        at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:88)
        at com.sun.proxy.$Proxy24.didClose(Unknown Source)
        at org.eclipse.lsp4e.DocumentContentSynchronizer.lambda$3(DocumentContentSynchronizer.java:153)
        at org.eclipse.lsp4e.DocumentContentSynchronizer$$Lambda$648/1763030548.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
        at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

```

**Comment by norru** [comment]
```
"org.eclipse.lsp4e.operations.folding.LSPFoldingReconciler" #109 daemon prio=1 os_prio=0 tid=0x0000000002436800 nid=0x6932 waiting for monitor entry [0x00007f1d5fbfa000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at org.eclipse.lsp4e.LanguageServerWrapper.start(LanguageServerWrapper.java:175)
        - waiting to lock <0x00000003c9a776d0> (a org.eclipse.lsp4e.LanguageServerWrapper)
        at org.eclipse.lsp4e.LanguageServerWrapper.getInitializedServer(LanguageServerWrapper.java:564)
        at org.eclipse.lsp4e.LanguageServerWrapper.getServerCapabilities(LanguageServerWrapper.java:596)
        at org.eclipse.lsp4e.LanguageServiceAccessor.lambda$11(LanguageServiceAccessor.java:482)
        at org.eclipse.lsp4e.LanguageServiceAccessor$$Lambda$501/1515167857.test(Unknown Source)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at org.eclipse.lsp4e.LanguageServiceAccessor.getLSPDocumentInfosFor(LanguageServiceAccessor.java:483)
        at org.eclipse.lsp4e.operations.folding.LSPFoldingReconcilingStrategy.reconcile(LSPFoldingReconcilingStrategy.java:121)
        at org.eclipse.lsp4e.operations.folding.LSPFoldingReconcilingStrategy.initialReconcile(LSPFoldingReconcilingStrategy.java:334)
        at org.eclipse.jface.text.reconciler.MonoReconciler.initialProcess(MonoReconciler.java:98)
        at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:177)
```

**Comment by norru** [comment]
```pkill rls``` unfreezes the IDE

**Comment by norru** [comment]
I haven't seen this for a while. Closing, will reopen if I see it again.

**Comment by norru** [comment]
I'm seeing this a lot, again.

I've attempted to post this as an issue in RLS https://github.com/rust-lang/rls/issues/1431

@alexheretic's comments seem to point to an integration issue rather than a problem with RLS itself.

**Comment by norru** [comment]
> RLS requires to be started from a cwd containing a Cargo.toml.

If this is the case, then there must be a mechanism to change the `cwd` before starting the server process stream in `LanguageServerWrapper.java`

Is it something that can be done cleanly?

**Comment by mickaelistria** [comment]
Thanks for your investigation and the Corrosion <-> RLS shuttle.

> multiple initialize

LSP4E decides when to start a new LS or not. It's expected that a LS supporting `workspaceFolders` capability can be reused for multiple projects. If the workspaceFolders capability is not set, then one LS instance is started for each project. It's an implementation choice made by LSP4E as the spec is very weak on that topic. But it seems to work for most existing LS. There is one and only one `initialize` request for each LS instance.

> RLS requires to be started from a cwd containing a Cargo.toml. What's the point of rootUri then?

That's also my thought, this requirement for RLS is not compatible with the idea of the spec IMO and should be improved in RLS to benefit to the LS and most adopters.

> RLS trying to set rootUri to /

It's something we implemented to allow hover in the .cargo files. Maybe it's not relevant anymore and this could be removed with better results (so could be the addition of the .cargo folder since LSP4E improved support for external files), but it's a bit hard to follow RLS on that topic as RLS seems to change a lot its assumptions and requirements without really evaluating conformance to spec or most adopters.
Maybe you can give a try to the following scenario: delete the `.cargo` project in the Eclipse IDE, and then start using your project and Ctrl+Click to navigate to libraries in `.cargo` folder. Those files would be open externally 

> If this is the case, then there must be a mechanism to change the cwd before starting the server process stream in LanguageServerWrapper.java

Unfortunately, there isn't such possibility in LSP4E at the moment, because all other supported LS can be started in a "stateless" way (independently of the project it's going to support) and then use the rootURI and similar to build the state.
I think RLS should try to build the same simple and efficient approach instead of piling a set of specific requirements that don't seem to provide added-value over conventions used by other LSs.

**Comment by norru** [comment]
> Unfortunately, there isn't such possibility in LSP4E at the moment, because all other supported LS can be started in a "stateless" way (independently of the project it's going to support) and then use the rootURI and similar to build the state.
> I think RLS should try to build the same simple and efficient approach instead of piling a set of specific requirements that don't seem to provide added-value over conventions used by other LSs.

If this is the case then we need to open a RLS issue for this and making the case for the RLS not actually following the LSP specs, by introducing unnecessary out-of-spec constraints. This is clearly unnecessary as there is a perfectly valid mechanism in-spec to obtain exactly the same result (`rootUri`)/

Again, my fear is that this is likely to be met by a "in VS Code it works this way so we're happy the way it is" which is likely to lead one of those soul-shattering and frustrating discussions on which side of the LSP things have to be implemented :(

**Comment by norru** [comment]
> Maybe you can give a try to the following scenario: delete the .cargo project in the Eclipse IDE, and then start using your project and Ctrl+Click to navigate to libraries in .cargo folder. Those files would be open externally

Tried this:

- if `/Cargo.toml` doesn't exist browsing to a dependency causes the IDE to freeze solid. `pkill rls` unfreezes the IDE but renders it unstable (freezes again shortly afterwards, freezes on shutdown, etc.)
- if '/Cargo.toml` does exist browsing to a dependency produces the following effect:
  - source file is opened in the IDE and syntax-colored as a Rust source file
  - **breakpoints** cannot be set
  - hovering/clicking has no **visible** effect, but RLS logs show that the `textDocument/definition` and `textDocument/hover` request are being issued and fulfilled (the response is being ignored by the editor).

**Comment by norru** [comment]
There is already a bug open for this: https://github.com/rust-lang/rls/issues/1135

**Comment by norru** [comment]
Updated workaround:

- create an empty `Cargo.toml` in the running directory of Eclipse Corrosion *and*
- create a `Cargo.toml` containing an empty `[workspace]` section in `~/.cargo`

**Comment by howaboutsynergy** [comment]
> create an empty Cargo.toml in the running directory of Eclipse Corrosion and

Either I didn't understand that*, or I just had to do this to fix it: `sudo touch /Cargo.toml`

\* I tried making cwd `/tmp` and `touch /tmp/Cargo.toml` then run `eclipse` (on Arch Linux: `eclipse-rust 4.11-1`) but that had no effect.

This also worked, got rid of one of the errors(they were twins):
> create a Cargo.toml containing an empty [workspace] section in ~/.cargo

**Comment by norru** [comment]
> sudo touch /Cargo.toml

This might fix Corrosion but work but makes terminal-run `cargo build` fail for most projects. `cargo build` complains about projects not being included in the root `[workspace]`

> create a Cargo.toml containing an empty [workspace] section in ~/.cargo

This works sometimes, not 100% (not understood the criteria)

All in all, peppering the workspace with random `Cargo.toml` files is a terrible and unreliable long term workaround.

We need a proper fix.

**Comment by norru** [comment]
@mickaelistria 

I'm currently investigating this and I'd appreciate some assistance if possible :)

Here's where I am at, still trying to make heads or tails.

- I can reproduce the problem in debug mode running Corrosion as an Eclipse application.
- I have removed the ```.cargo``` project from the target workspace and commented out the code that automatically creates it.
- With the ```.cargo``` project removed, navigation (go to declaration) seems to work correctly for both project files and external libraries files
- navigating to a symbol in a source library file from the```.cargo``` opens it in the "rust" editor as expected. This also happens when a file in the "rustsrc" standard lib is opened, as expected.
- it is **not** possible to place a breakpoint in such source files (although the editor seems to be associated to a Rust source file.
- a new `rls` instance appears to be created for every ```.cargo``` crate when a source file it contains is open, which is acceptable. Every time a new RLS is instantiated, the 
> could not find `Cargo.toml` in `/` or any parent directory appears in a popup
which is the bit I'm currently interested to know the root cause of.
- When the last editor belonging to a ```.cargo``` crate closes, the associated `rls` appears to be killed and the following exception is logged multiple times:

```
Jul 27, 2019 9:09:19 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint notify
WARNING: Failed to send notification message.
java.lang.NullPointerException
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$1(LanguageServerWrapper.java:216)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.notify(RemoteEndpoint.java:126)
	at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:88)
	at com.sun.proxy.$Proxy24.didClose(Unknown Source)
	at org.eclipse.lsp4e.DocumentContentSynchronizer.lambda$3(DocumentContentSynchronizer.java:157)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```
Your thoughts?

**Comment by norru** [comment]
I think I have moved a bit further. There is this piece of code in `LanguageServerWrapper.java` in LSP4E:

```java
			if (project != null && project.exists()) {
				rootURI = LSPEclipseUtils.toUri(this.initialProject);
				initParams.setRootUri(rootURI.toString());
				initParams.setRootPath(rootURI.getPath());
			} else {
				initParams.setRootUri(LSPEclipseUtils.toUri(new File("/")).toString()); //$NON-NLS-1$
			}
```

Basically, if I understood correctly is "determine the cwd (to pass to initialization) of the LS by getting the root of the current project. If there is no current project, then use the root '/'"

This happens when navigating out from a Cargo Eclipse project into a library in `.cargo'.

In RLS this is clearly a problem because '/' is not a valid Rust/Cargo root project.

LSP4E needs a better way to figure out what the root of the project is. The best way I can think of is to keep track of the path of the file being open in the editor and determine the default root by crawling up the file system from the original file path until a `Cargo.toml` is found.

At the moment I don't see this piece of information being exposed anywhere.

**Comment by norru** [comment]
@mickaelistria

So, I do have a fix for this. Unfortunately it's in LSP4E which is hosted on ```git.eclipse.org``` and I have no idea how to submit a pull request.

It does seem to work, but ```.cargo``` needs to be in the workspace for breakpoints to work. I have also tracked that down to a known issue whereas any file opened by TM4E from a non-workspace file has no ```IResource``` attached to it.

I believe that if that one gets fixed we wouldn't need the ```.cargo``` folder at all.

**Comment by mickaelistria** [comment]
If you think removing `.cargo` is better, we can remove it.

So the only real remaining issue is that we cannot put breakpoints on external files, isn't it?

To submit a change to a project on Eclipse's Git, please use Gerrit: https://wiki.eclipse.org/Gerrit . Basically, upload you .ssh key to https://git.eclipse.org/r/#/settings, add remotes you can see on https://git.eclipse.org/r/#/admin/projects/lsp4e/lsp4e and then `git push gerrit HEAD:refs/for/master` to create a new review.

**Comment by mickaelistria** [comment]
About LSP4E, before sending patches, please make sure you have a Bugzilla report for each of the issue you want to fix (unless it's really trivial), and make your commits start with "Bug 123345 - <...>" for easier reference. Also, please keep the commits separated: your branch tries to fix a lot of things that are separate issue, separating commits/reviews will ease progress.

**Comment by norru** [comment]
> If you think removing .cargo is better, we can remove it.
> So the only real remaining issue is that we cannot put breakpoints on external files, isn't it?

As things stand, ```.cargo``` cannot be removed because of the ```IResource``` thing above, but yes, I believe this workaround needs to be removed eventually. It's untidy, and confusing at best (I've read comments on Corrosion's issue tracker).

Once the ```IResource``` issue gets fixed, I believe there is no reason to keep the ```.cargo``` workaround, but we'll have to confirm.

> Also, please keep the commits separated: your branch tries to fix a lot of things that are separate issue, separating commits/reviews will ease progress.

Can do, the pull request above was mostly for you to have a look at. No point in posting "proper" patches if the approach taken is unreasonable. I was in fact worried that other Language Servers be affected by the change above.

Although, the Gerrit workflow does appear to be painful, at least on paper.

**Comment by akurtakov** [comment]
Would you please give pointer to the IResource issue? I didn't manage to figure it out from comments here.

**Comment by norru** [comment]
https://github.com/itadinanta/lsp4e/pull/1/commits/72d7f51783b8a967564fb0c93d7c12f0df7beb40#diff-88f116cd37a224f702a102875c22fdc8

org.eclipse.lsp4e.debug/src/org/eclipse/lsp4e/debug/breakpoints/DSPBreakpointAdapter.java

line 28

```
IResource resource = textEditor.getEditorInput().getAdapter(IResource.class);
```

```resource``` is null if the textEditor refers to a file which is not in the workspace (such as an external source). As such, it's impossible to place a breakpoint on. In fact, it does NPE a few lines below while trying to remove the breakpoint in first place (but that's an easy fix).

**Comment by mickaelistria** [comment]
> No point in posting "proper" patches if the approach taken is unreasonable.

I'm really busy with many things and can't really spend time on reviewing or evaluating "improper" patches. Focused and smaller patches/bugs are more actionable.
And there is nothing really unreasonable when it's fixing an actual issue ;) In worst case, it's a first iteration; in best case, it's a fix.

> the Gerrit workflow does appear to be painful, at least on paper.

It seems painful because it's different from the PR one, but in the end, it works the same except you start the PR when pushing the commit (by pushing to refs/for/master) instead of creating it from some UI. The the 1 commit <-> 1 review model is really much more efficient when familiar with it.
So that's worth trying ;)

**Comment by norru** [comment]
> The the 1 commit <-> 1 review model is really much more efficient when familiar with it.
So that's worth trying ;)

Do I need to squash beforehand?

**Comment by mickaelistria** [comment]
> Do I need to squash beforehand?

Yes, unless the commits make sense separately (ie they are meaningful design iterations), it's better to have them in a single commit/review; if it's a more complex and iterative design, one commit/review per step is also good.

**Comment by norru** [comment]
Process looks tedious. I'll give it a go just to see how it goes but I am unlikely to be able to cope with the overhead for too long.

**Comment by norru** [comment]
I'm trying to wrap my head around Gerrit and I am struggling to figure out a workflow which doesn't lose history (ie preserve the history that would be otherwise lost with "--amend"). Any clues?

**Comment by mickaelistria** [comment]
:D I'm hearing you as an echo of myself some years ago when I have similar opinion.

The truth is that both the project and the reviewer don't really need to know nor store your work history, it needs to keep an history of meaningful commits for easier tracking in the future, whether you made a commit to fix a typo or a minor bug in your patch is actually more noise than value from project perspective (by project I mean current and future contributors). You're free to push multiple successive commits with Gerrit if that makes sense for the reviewer and the project, they will just be multiple Gerrit reviews (instead of a single PR in GitHub world).
For your first experiment, just accept losing your local fine grain history and embrace the Gerrit way even if you don't like it at first, and then you'll progressively adhere to Gerrit workflow and realize how better it is than GitHub ones ;)

**Comment by norru** [comment]
ttps://bugs.eclipse.org/bugs/show_bug.cgi?id=549758

**Comment by norru** [comment]
@mickaelistria 
```
nico@lenovo-devuan:~/Projects/3rdParty$ git clone git://git.eclipse.org/gitroot/lsp4e/lsp4e lsp4e_copy
Cloning into 'lsp4e_copy'...
fatal: remote error: access denied or repository not exported: /gitroot/lsp4e/lsp4e
```

I was able to clone/fetch a few days ago.

**Comment by mickaelistria** [comment]
Thanks for the reports and the Git attempt.
Unfortunately, defaut Git is in trouble today, I suggest you use the URLs fir Gerrit instead, that seems to still work well, ar least the SSH one. See https://git.eclipse.org/r/#/admin/projects/lsp4e/lsp4e for the remote URLs

**Comment by norru** [comment]
Right, I did manage to fetch from ssh://git.eclipse.org:29418/lsp4e/lsp4e, I'll see if I can attempt a Gerrit push in the next few days.

Wish me luck :D

**Comment by mickaelistria** [comment]
🍀
If you get other questions/issues.with Gerrit,feel free to ask for assistance (on the LSP4E bugs preferably)

**Comment by norru** [comment]
Fixed in LSP4E, at least partially.

Caveat: an empty `Cargo.toml` still needs to exist in `.cargo`.

`.cargo` project can be removed once the LSP4E Debug Adapter is fixed. At the moment, without the `.cargo` project, a breakpoint cannot be placed in library code.
